Struct generation improvements

SlayerOfTheBad opened the pull request:
This PR includes multiple improvements to virtue regarding the generation of Structs and Enums:
1. It makes it possible for structs to have unnamed fields, or be a ZST (`struct StructName;`).
2. It makes it possible for Enums to be generated, and all types of values therein.
3. It makes it possible to generate Enums and Structs outside of a module.

Points 2 and 3 have been previously raised in https://github.com/bincode-org/virtue/issues/23.
Point 1 is just something I ran into personally.

SlayerOfTheBad committed:
Make it possible to generate unnamed and ZST structs

Adds the possibility to make ZST structs and structs with unnamed
fields.
Backwards compability is insured by defaulting to `StructType::Named`,
which causes identical behaviour to before.

SlayerOfTheBad committed:
Add generate enum capability

Adds an enum generator, based on the struct generator.

SlayerOfTheBad committed:
Make it possible to generate structs at the root

Previously, structs could only be generated inside a new module.
Sometimes, it is desirable to define a new struct in the same module as
the struct that is being derived on.
By dupicating the `generate_struct` function from `GenerateMod` to to
`Generator`, this is easily achieved.

VictorKoenders commented:
Would you mind adding examples that will also validate the generated code? An example can be found [here](https://github.com/bincode-org/virtue/blob/trunk/src/generate/impl_for.rs#L83-L97)

SlayerOfTheBad committed:
Add doc comment tests

Add testcases for the struct and enum generators.

SlayerOfTheBad commented:
I have added doc comment tests that I believe cover all of the added functionality. Please let me know if I missed anything!

SlayerOfTheBad committed:
Improve function names

A Tuple struct is called a Tuple struct, not an unnamed field struct, so
it makes more sense for the function to be called `make_tuple`, in line
with `make_ZST`.

VictorKoenders commented:
Thanks!
